adapters/tlsio_openssl.c

```diff
@@ -822,10 +831,429 @@ static void on_underlying_io_bytes_received(void* context, const unsigned char*
+    {
+        if (str[start] < '0' || str[start] > '9')
+            return -1;
+
+        result = (result * 10) + (str[start] - '0');
+    }
+
+    return result;
+}
+
+static time_t crl_invalid_after(X509_CRL *crl)
+{
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER < 0x20000000L)
+    const ASN1_TIME *at = X509_CRL_get0_nextUpdate(crl);
+#else
+    ASN1_TIME *at = crl->crl->nextUpdate;
+#endif
+
+    ASN1_GENERALIZEDTIME *gt = ASN1_TIME_to_generalizedtime(at, NULL);
+    if (!gt)
+    {
+        LogError("crl could not find field\n");
+        return 0;
+    }
+    // LogInfo("crl data %s\n", (char*)gt->data);
+
+    // "20181011181119Z"
+    int success = gt->length >= 14 ? 1 : 0;
+    struct tm tm = { 0, };
+
+    if (success)
+    {
+        tm.tm_year = atoin((char*)gt->data, 0, 4) - 1900;
+        tm.tm_mon = atoin((char*)gt->data, 4, 2) - 1;
+        tm.tm_mday = atoin((char*)gt->data, 6, 2);
+        tm.tm_hour = atoin((char*)gt->data, 8, 2);
+        tm.tm_min = atoin((char*)gt->data, 10, 2);
+        tm.tm_sec = atoin((char*)gt->data, 12, 2);
+
+        success = (tm.tm_year > 100 && tm.tm_mon >= 0 && tm.tm_mday > 0 &&
+            tm.tm_hour >= 0 && tm.tm_min >= 0 && tm.tm_sec >= 0);
+    }
+
+    ASN1_GENERALIZEDTIME_free(gt);
+    if (!success)
+    {
+        return 0;
+    }
+
+    // calculates the time since epoch
+    return mktime(&tm);
+}
+
+static const char *get_dp_url(DIST_POINT *dp)
+{
+    GENERAL_NAMES *gens;
+    GENERAL_NAME *gen;
+    int i, gtype;
+    ASN1_STRING *uri;
+
+    if (!dp->distpoint || dp->distpoint->type != 0)
+    {
+        return NULL;
+    }
+
+    gens = dp->distpoint->name.fullname;
+
+    for (i = 0; i < sk_GENERAL_NAME_num(gens); i++)
+    {
+        gen = sk_GENERAL_NAME_value(gens, i);
+        uri = GENERAL_NAME_get0_value(gen, &gtype);
+
+        if (gtype == GEN_URI && ASN1_STRING_length(uri) > 6)
+        {
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER < 0x20000000L)
+            char *uptr = (char *)ASN1_STRING_get0_data(uri);
+#else
+            char *uptr = (char *)ASN1_STRING_data(uri);
+#endif
+            if (!strncmp(uptr, "http://", 7))
+            {
+                return uptr;
+            }
+        }
+    }
+
+    return NULL;
+}
+
+static X509_CRL *load_crl_crldp(X509 *cert, const char* suffix, STACK_OF(DIST_POINT) *crldp)
+{
+    int i;
+    X509_CRL *crl = NULL;
+    char buf[256];
+    time_t now = time(NULL);
+
+    char* prefix = NULL;
+    if( NULL == (prefix = getenv("TMP")) &&
+        NULL == (prefix = getenv("TEMP")) &&
+        NULL == (prefix = getenv("TMPDIR")))
+    {
+        prefix = ".";
+    }
+
+    // we need the issuer hash to find the file on disk
+    X509_NAME *issuer_cert = X509_get_issuer_name(cert);
+    unsigned long hash = X509_NAME_hash(issuer_cert);
+
+    // try to read from file
+    for (i = 0; i < 10; i++)
+    {
+        sprintf(buf, "%s/%08lx.%s.%d", prefix, hash, suffix, i);
```

Comment:
The file name here is kind of generic. We can not exclude that this file is created by us, so when you delete it later (if it is expired), you might delete the file created by others. Having something like "SpeechSDK" in the file name will decrease the name duplication.

Reply:
not really, the name is a hash value so collisions should be rare already.